### ISSUE TITLE
Build errors on Ubuntu 22.04

### ISSUE THREAD

**Issue by FrankRay78** [ondrejbudai/hidviz#34]
I followed the instructions exactly here: https://github.com/ondrejbudai/hidviz, and the following build errors occurred:

```
frankray@frank-ubuntu-virtual-machine:~/Repos/hidviz/build$ make -j$(nproc)
[  2%] Running cpp protocol buffer compiler on src/Buffer.proto
[  8%] Building CXX object libhidx/libhidx_server/CMakeFiles/hidx_server.dir/Buffer.pb.cc.o
[  5%] Building CXX object libhidx/libhidx_server/CMakeFiles/hidx_server.dir/src/Server.cc.o
[ 11%] Building CXX object libhidx/libhidx_server/CMakeFiles/hidx_server.dir/src/Utils.cc.o
[ 13%] Linking CXX static library libhidx_server.a
[ 13%] Built target hidx_server
[ 22%] Building CXX object libhidx/libhidx/CMakeFiles/hidx.dir/src/LibHidx.cc.o
[ 22%] Building CXX object libhidx/libhidx/CMakeFiles/hidx.dir/src/Device.cc.o
[ 22%] Building CXX object libhidx/libhidx_server_daemon/CMakeFiles/libhidx_server_daemon.dir/src/main.cc.o
[ 25%] Building CXX object libhidx/libhidx/CMakeFiles/hidx.dir/src/InterfaceHandle.cc.o
[ 27%] Building CXX object libhidx/libhidx/CMakeFiles/hidx.dir/src/Interface.cc.o
[ 30%] Building CXX object libhidx/libhidx/CMakeFiles/hidx.dir/src/Parser.cc.o
[ 33%] Building CXX object libhidx/libhidx/CMakeFiles/hidx.dir/src/hid/Collection.cc.o
[ 36%] Building CXX object libhidx/libhidx/CMakeFiles/hidx.dir/src/hid/Item.cc.o
[ 38%] Building CXX object libhidx/libhidx/CMakeFiles/hidx.dir/src/hid/Control.cc.o
[ 41%] Building CXX object libhidx/libhidx/CMakeFiles/hidx.dir/src/Usages.cc.o
In file included from /home/frankray/Repos/hidviz/libhidx/libhidx/src/Interface.cc:21:
/home/frankray/Repos/hidviz/libhidx/libhidx/include/libhidx/Interface.hh:76:14: error: ‘shared_ptr’ in namespace ‘std’ does not name a template type
   76 |         std::shared_ptr<InterfaceHandle> getHandle();
      |              ^~~~~~~~~~
/home/frankray/Repos/hidviz/libhidx/libhidx/include/libhidx/Interface.hh:29:1: note: ‘std::shared_ptr’ is defined in header ‘<memory>’; did you forget to ‘#include <memory>’?
   28 | #include <string>
  +++ |+#include <memory>
   29 | 
/home/frankray/Repos/hidviz/libhidx/libhidx/include/libhidx/Interface.hh:120:14: error: ‘weak_ptr’ in namespace ‘std’ does not name a template type
  120 |         std::weak_ptr<InterfaceHandle> m_handle;
      |              ^~~~~~~~
/home/frankray/Repos/hidviz/libhidx/libhidx/include/libhidx/Interface.hh:120:9: note: ‘std::weak_ptr’ is defined in header ‘<memory>’; did you forget to ‘#include <memory>’?
  120 |         std::weak_ptr<InterfaceHandle> m_handle;
      |         ^~~
[ 44%] Building CXX object libhidx/libhidx/CMakeFiles/hidx.dir/src/hid/Usage.cc.o
[ 47%] Building CXX object libhidx/libhidx/CMakeFiles/hidx.dir/src/Connector.cc.o
/home/frankray/Repos/hidviz/libhidx/libhidx/src/Interface.cc: In member function ‘void libhidx::Interface::parseHidReportDesc()’:
/home/frankray/Repos/hidviz/libhidx/libhidx/src/Interface.cc:84:23: error: ‘getHandle’ was not declared in this scope; did you mean ‘handle’?
   84 |         auto handle = getHandle();
      |                       ^~~~~~~~~
      |                       handle
/home/frankray/Repos/hidviz/libhidx/libhidx/src/Interface.cc:95:114: error: no matching function for call to ‘libhidx::ConnectionException::ConnectionException(<brace-enclosed initializer list>)’
   95 |             throw ConnectionException{data.first, "Libusb control transfer failed: " + std::to_string(data.first)};
      |                                                                                                                  ^
In file included from /home/frankray/Repos/hidviz/libhidx/libhidx/src/Interface.cc:24:
/home/frankray/Repos/hidviz/libhidx/libhidx/include/libhidx/InterfaceHandle.hh:37:18: note: candidate: ‘libhidx::ConnectionException::ConnectionException(int, const string&)’
   37 |         explicit ConnectionException(int ev, const std::string& err) : std::system_error{ev, std::generic_category(),  err}{}
      |                  ^~~~~~~~~~~~~~~~~~~
/home/frankray/Repos/hidviz/libhidx/libhidx/include/libhidx/InterfaceHandle.hh:37:18: note:   conversion of argument 1 would be ill-formed:
/home/frankray/Repos/hidviz/libhidx/libhidx/include/libhidx/InterfaceHandle.hh:35:11: note: candidate: ‘libhidx::ConnectionException::ConnectionException(const libhidx::ConnectionException&)’
   35 |     class ConnectionException : public std::system_error {
      |           ^~~~~~~~~~~~~~~~~~~
/home/frankray/Repos/hidviz/libhidx/libhidx/include/libhidx/InterfaceHandle.hh:35:11: note:   candidate expects 1 argument, 2 provided
/home/frankray/Repos/hidviz/libhidx/libhidx/include/libhidx/InterfaceHandle.hh:35:11: note: candidate: ‘libhidx::ConnectionException::ConnectionException(libhidx::ConnectionException&&)’
/home/frankray/Repos/hidviz/libhidx/libhidx/include/libhidx/InterfaceHandle.hh:35:11: note:   candidate expects 1 argument, 2 provided
/home/frankray/Repos/hidviz/libhidx/libhidx/src/Interface.cc: In member function ‘void libhidx::Interface::runner()’:
/home/frankray/Repos/hidviz/libhidx/libhidx/src/Interface.cc:149:23: error: ‘getHandle’ was not declared in this scope; did you mean ‘handle’?
  149 |         auto handle = getHandle();
      |                       ^~~~~~~~~
      |                       handle
/home/frankray/Repos/hidviz/libhidx/libhidx/src/Interface.cc:161:103: error: no matching function for call to ‘std::vector<unsigned char>::vector(<brace-enclosed initializer list>)’
  161 |                 std::vector<unsigned char> data{std::begin(response.data()), std::end(response.data())};
      |                                                                                                       ^
In file included from /usr/include/c++/11/vector:67,
                 from /usr/include/c++/11/functional:62,
                 from /home/frankray/Repos/hidviz/libhidx/libhidx/include/libhidx/Interface.hh:26,
                 from /home/frankray/Repos/hidviz/libhidx/libhidx/src/Interface.cc:21:
/usr/include/c++/11/bits/stl_vector.h:653:9: note: candidate: ‘template<class _InputIterator, class> std::vector<_Tp, _Alloc>::vector(_InputIterator, _InputIterator, const allocator_type&) [with _InputIterator = _InputIterator; <template-parameter-2-2> = <template-parameter-1-2>; _Tp = unsigned char; _Alloc = std::allocator<unsigned char>]’
  653 |         vector(_InputIterator __first, _InputIterator __last,
      |         ^~~~~~
/usr/include/c++/11/bits/stl_vector.h:653:9: note:   template argument deduction/substitution failed:
/usr/include/c++/11/bits/stl_vector.h:625:7: note: candidate: ‘std::vector<_Tp, _Alloc>::vector(std::initializer_list<_Tp>, const allocator_type&) [with _Tp = unsigned char; _Alloc = std::allocator<unsigned char>; std::vector<_Tp, _Alloc>::allocator_type = std::allocator<unsigned char>]’
  625 |       vector(initializer_list<value_type> __l,
      |       ^~~~~~
/usr/include/c++/11/bits/stl_vector.h:625:7: note:   conversion of argument 1 would be ill-formed:
/usr/include/c++/11/bits/stl_vector.h:607:7: note: candidate: ‘std::vector<_Tp, _Alloc>::vector(std::vector<_Tp, _Alloc>&&, const allocator_type&) [with _Tp = unsigned char; _Alloc = std::allocator<unsigned char>; std::vector<_Tp, _Alloc>::allocator_type = std::allocator<unsigned char>]’
  607 |       vector(vector&& __rv, const allocator_type& __m)
      |       ^~~~~~
/usr/include/c++/11/bits/stl_vector.h:607:7: note:   conversion of argument 1 would be ill-formed:
/usr/include/c++/11/bits/stl_vector.h:589:7: note: candidate: ‘std::vector<_Tp, _Alloc>::vector(std::vector<_Tp, _Alloc>&&, const allocator_type&, std::false_type) [with _Tp = unsigned char; _Alloc = std::allocator<unsigned char>; std::vector<_Tp, _Alloc>::allocator_type = std::allocator<unsigned char>; std::false_type = std::integral_constant<bool, false>]’
  589 |       vector(vector&& __rv, const allocator_type& __m, false_type)
      |       ^~~~~~
/usr/include/c++/11/bits/stl_vector.h:589:7: note:   candidate expects 3 arguments, 2 provided
/usr/include/c++/11/bits/stl_vector.h:585:7: note: candidate: ‘std::vector<_Tp, _Alloc>::vector(std::vector<_Tp, _Alloc>&&, const allocator_type&, std::true_type) [with _Tp = unsigned char; _Alloc = std::allocator<unsigned char>; std::vector<_Tp, _Alloc>::allocator_type = std::allocator<unsigned char>; std::true_type = std::integral_constant<bool, true>]’
  585 |       vector(vector&& __rv, const allocator_type& __m, true_type) noexcept
      |       ^~~~~~
/usr/include/c++/11/bits/stl_vector.h:585:7: note:   candidate expects 3 arguments, 2 provided
/usr/include/c++/11/bits/stl_vector.h:575:7: note: candidate: ‘std::vector<_Tp, _Alloc>::vector(const std::vector<_Tp, _Alloc>&, const allocator_type&) [with _Tp = unsigned char; _Alloc = std::allocator<unsigned char>; std::vector<_Tp, _Alloc>::allocator_type = std::allocator<unsigned char>]’
  575 |       vector(const vector& __x, const allocator_type& __a)
      |       ^~~~~~
/usr/include/c++/11/bits/stl_vector.h:575:7: note:   conversion of argument 1 would be ill-formed:
/usr/include/c++/11/bits/stl_vector.h:572:7: note: candidate: ‘std::vector<_Tp, _Alloc>::vector(std::vector<_Tp, _Alloc>&&) [with _Tp = unsigned char; _Alloc = std::allocator<unsigned char>]’
  572 |       vector(vector&&) noexcept = default;
      |       ^~~~~~
/usr/include/c++/11/bits/stl_vector.h:572:7: note:   candidate expects 1 argument, 2 provided
/usr/include/c++/11/bits/stl_vector.h:553:7: note: candidate: ‘std::vector<_Tp, _Alloc>::vector(const std::vector<_Tp, _Alloc>&) [with _Tp = unsigned char; _Alloc = std::allocator<unsigned char>]’
  553 |       vector(const vector& __x)
      |       ^~~~~~
/usr/include/c++/11/bits/stl_vector.h:553:7: note:   candidate expects 1 argument, 2 provided
/usr/include/c++/11/bits/stl_vector.h:522:7: note: candidate: ‘std::vector<_Tp, _Alloc>::vector(std::vector<_Tp, _Alloc>::size_type, const value_type&, const allocator_type&) [with _Tp = unsigned char; _Alloc = std::allocator<unsigned char>; std::vector<_Tp, _Alloc>::size_type = long unsigned int; std::vector<_Tp, _Alloc>::value_type = unsigned char; std::vector<_Tp, _Alloc>::allocator_type = std::allocator<unsigned char>]’
  522 |       vector(size_type __n, const value_type& __value,
      |       ^~~~~~
/usr/include/c++/11/bits/stl_vector.h:522:7: note:   conversion of argument 1 would be ill-formed:
/usr/include/c++/11/bits/stl_vector.h:510:7: note: candidate: ‘std::vector<_Tp, _Alloc>::vector(std::vector<_Tp, _Alloc>::size_type, const allocator_type&) [with _Tp = unsigned char; _Alloc = std::allocator<unsigned char>; std::vector<_Tp, _Alloc>::size_type = long unsigned int; std::vector<_Tp, _Alloc>::allocator_type = std::allocator<unsigned char>]’
  510 |       vector(size_type __n, const allocator_type& __a = allocator_type())
      |       ^~~~~~
/usr/include/c++/11/bits/stl_vector.h:510:7: note:   conversion of argument 1 would be ill-formed:
/usr/include/c++/11/bits/stl_vector.h:497:7: note: candidate: ‘std::vector<_Tp, _Alloc>::vector(const allocator_type&) [with _Tp = unsigned char; _Alloc = std::allocator<unsigned char>; std::vector<_Tp, _Alloc>::allocator_type = std::allocator<unsigned char>]’
  497 |       vector(const allocator_type& __a) _GLIBCXX_NOEXCEPT
      |       ^~~~~~
/usr/include/c++/11/bits/stl_vector.h:497:7: note:   candidate expects 1 argument, 2 provided
/usr/include/c++/11/bits/stl_vector.h:487:7: note: candidate: ‘std::vector<_Tp, _Alloc>::vector() [with _Tp = unsigned char; _Alloc = std::allocator<unsigned char>]’
  487 |       vector() = default;
      |       ^~~~~~
/usr/include/c++/11/bits/stl_vector.h:487:7: note:   candidate expects 0 arguments, 2 provided
/home/frankray/Repos/hidviz/libhidx/libhidx/src/Interface.cc: At global scope:
/home/frankray/Repos/hidviz/libhidx/libhidx/src/Interface.cc:176:38: error: no declaration matches ‘std::shared_ptr<libhidx::InterfaceHandle> libhidx::Interface::getHandle()’
  176 |     std::shared_ptr<InterfaceHandle> Interface::getHandle() {
      |                                      ^~~~~~~~~
/home/frankray/Repos/hidviz/libhidx/libhidx/src/Interface.cc:176:38: note: no functions named ‘std::shared_ptr<libhidx::InterfaceHandle> libhidx::Interface::getHandle()’
In file included from /home/frankray/Repos/hidviz/libhidx/libhidx/src/Interface.cc:21:
/home/frankray/Repos/hidviz/libhidx/libhidx/include/libhidx/Interface.hh:45:11: note: ‘class libhidx::Interface’ defined here
   45 |     class Interface {
      |           ^~~~~~~~~
/home/frankray/Repos/hidviz/libhidx/libhidx/src/Interface.cc: In member function ‘void libhidx::Interface::sendOutputReport(const std::vector<unsigned char>&)’:
/home/frankray/Repos/hidviz/libhidx/libhidx/src/Interface.cc:253:23: error: ‘getHandle’ was not declared in this scope; did you mean ‘handle’?
  253 |         auto handle = getHandle();
      |                       ^~~~~~~~~
      |                       handle
[ 50%] Linking CXX executable libhidx_server_daemon
[ 50%] Built target libhidx_server_daemon
make[2]: *** [libhidx/libhidx/CMakeFiles/hidx.dir/build.make:118: libhidx/libhidx/CMakeFiles/hidx.dir/src/Interface.cc.o] Error 1
make[2]: *** Waiting for unfinished jobs....
make[1]: *** [CMakeFiles/Makefile2:222: libhidx/libhidx/CMakeFiles/hidx.dir/all] Error 2
make: *** [Makefile:136: all] Error 2
```

I checked my version of gcc and it is:

```
frankray@frank-ubuntu-virtual-machine:~/Repos/hidviz/build$ gcc --version
gcc (Ubuntu 11.3.0-1ubuntu1~22.04) 11.3.0
Copyright (C) 2021 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

**Comment by ldab** [comment]
Try to checkout the submodule `libhidx` to master, it fixed for me.

**Comment by ondrejbudai** [comment]
Fixed in 0.2
